File: src/sugtao4423/lod/tweetlistview/TweetListView.java
package sugtao4423.lod.tweetlistview;

import android.content.Context;
import android.graphics.Rect;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.Adapter;
import android.support.v7.widget.RecyclerView.ItemDecoration;
import android.util.AttributeSet;
import android.view.View;
import sugtao4423.lod.App;
import sugtao4423.lod.R;
import twitter4j.Status;

public class TweetListView extends RecyclerView{

	private LinearLayoutManager llm;

	public TweetListView(Context context){
		super(context);
		setVerticalScrollBarEnabled(true);
		addItemDecoration(new DividerItemDecoration(context, DividerItemDecoration.VERTICAL));
		addItemDecoration(new AlternatelyColor(context));
		llm = new LinearLayoutManager(context);
		setLayoutManager(llm);
	}

	public TweetListView(Context context, AttributeSet attrs){
		super(context, attrs);
		setVerticalScrollBarEnabled(true);
		addItemDecoration(new DividerItemDecoration(context, DividerItemDecoration.VERTICAL));
		addItemDecoration(new AlternatelyColor(context));
		llm = new LinearLayoutManager(context);
		setLayoutManager(llm);
	}

	public LinearLayoutManager getLinearLayoutManager(){
		return llm;
	}

}

class AlternatelyColor extends ItemDecoration{

	private App app;

	public AlternatelyColor(Context context){
		this.app = (App)context.getApplicationContext();
	}

	@Override
	public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state){
		super.getItemOffsets(outRect, view, parent, state);
		int pos = parent.getChildAdapterPosition(view);
		Adapter<?> adapter = parent.getAdapter();
		if(adapter instanceof TweetListAdapter){
			Status item = ((TweetListAdapter)adapter).getItem(pos);
			if(item == null)
				return;
			if(item.isRetweetedByMe()){
				view.setBackgroundResource(R.drawable.retweeted_by_me);
			}else if(item.isRetweet()){
				view.setBackgroundResource(R.drawable.retweet);
			}else if(item.getUser().getScreenName().equals(app.getMyScreenName())){
				view.setBackgroundResource(R.drawable.same_my_screenname);
			}else if(app.getMentionPattern().matcher(item.getText()).find()){
				view.setBackgroundResource(R.drawable.mention);
			}else{
				setAlternately(pos, view);
			}
		}else if(adapter instanceof TweetListUserAdapter){
			setAlternately(pos, view);
		}
	}

	public void setAlternately(int pos, View view){
		view.setBackgroundResource(pos % 2 == 0 ? R.drawable.position0 : R.drawable.position1);
	}

}

File: app/src/main/java/sugtao4423/lod/tweetlistview/TweetListView.kt
package sugtao4423.lod.tweetlistview

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Rect
import android.util.AttributeSet
import android.view.View
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import sugtao4423.lod.App
import sugtao4423.lod.R

class TweetListView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : RecyclerView(context, attrs, defStyle) {

    val linearLayoutManager: LinearLayoutManager

    init {
        isVerticalScrollBarEnabled = true
        addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        addItemDecoration(AlternatelyColor(context))
        linearLayoutManager = LinearLayoutManager(context)
        layoutManager = linearLayoutManager
    }

    inner class AlternatelyColor(context: Context) : ItemDecoration() {

        private val app = context.applicationContext as App

        @SuppressLint("ResourceType")
        override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: State) {
            super.getItemOffsets(outRect, view, parent, state)
            val pos = parent.getChildAdapterPosition(view)
            if (pos < 0) {
                return
            }
            val adapter = parent.adapter
            if (adapter is TweetListAdapter) {
                val item = adapter.data[pos]
                when {
                    item.isRetweetedByMe -> view.setBackgroundResource(R.xml.retweeted_by_me)
                    item.isRetweet -> view.setBackgroundResource(R.xml.retweet)
                    item.user.screenName == app.account.screenName -> view.setBackgroundResource(R.xml.same_my_screenname)
                    app.mentionPattern.matcher(item.text).find() -> view.setBackgroundResource(R.xml.mention)
                    else -> setAlternately(pos, view)
                }
            } else if (adapter is TweetListUserAdapter) {
                setAlternately(pos, view)
            }
        }

        @SuppressLint("ResourceType")
        private fun setAlternately(pos: Int, view: View) {
            view.setBackgroundResource(if (pos % 2 == 0) R.xml.position0 else R.xml.position1)
        }

    }

}
